{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene body counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "import collections\n",
    "import matplotlib.pyplot as plt\n",
    "import gzip\n",
    "import loompy\n",
    "# import scipy.sparse as sparse\n",
    "import urllib.request\n",
    "import pybedtools\n",
    "from pybedtools import BedTool\n",
    "import warnings\n",
    "from sklearn.neighbors import NearestNeighbors\n",
    "from matplotlib.collections import LineCollection\n",
    "\n",
    "sys.path.append('/home/camiel/chromograph/')\n",
    "# from chromograph.plotting.UMI_plot import UMI_plot\n",
    "import chromograph\n",
    "from chromograph.peak_calling.utils import *\n",
    "\n",
    "from umap import UMAP\n",
    "import sklearn.metrics\n",
    "from scipy.spatial import distance\n",
    "import community\n",
    "import networkx as nx\n",
    "from scipy import sparse\n",
    "from typing import *\n",
    "\n",
    "import logging\n",
    "logger = logging.getLogger()\n",
    "logging.basicConfig(\n",
    "    format='%(asctime)s %(levelname)-8s %(message)s',\n",
    "    level=logging.INFO,\n",
    "    datefmt='%H:%M:%S')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Import path to the relevant 10X reference dataset\n",
    "\n",
    "ref = '/data/ref/cellranger-atac/refdata-cellranger-atac-mm10-1.2.0'\n",
    "indir = '/data/proj/scATAC/chromograph/mouse_test2/'\n",
    "f = os.path.join(indir, '10X_test_10kb.loom')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Connect to loompy session\n",
    "ds = loompy.connect(f)\n",
    "logging.info(ds.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "76191\n",
      "chr1\t3073252\t3074322\tRP23-271O17.1\t.\t+\tTEC\n",
      " chr1\t3102015\t3102125\tGm26206\t.\t+\tsnRNA\n",
      " chr1\t3214481\t3671498\tXkr4\t.\t-\tprotein_coding\n",
      " chr1\t3252756\t3253236\tRP23-317L18.1\t.\t+\tprocessed_pseudogene\n",
      " chr1\t3365730\t3368549\tRP23-317L18.4\t.\t-\tTEC\n",
      " chr1\t3375555\t3377788\tRP23-317L18.3\t.\t-\tTEC\n",
      " chr1\t3464976\t3467285\tRP23-115I1.6\t.\t-\tTEC\n",
      " chr1\t3466586\t3513553\tRP23-115I1.1\t.\t+\tantisense\n",
      " chr1\t3512450\t3514507\tRP23-115I1.5\t.\t-\tTEC\n",
      " chr1\t3531794\t3532720\tRP23-115I1.2\t.\t+\tprocessed_pseudogene\n",
      " "
     ]
    }
   ],
   "source": [
    "genes = BedTool(os.path.join(ref, 'regions', 'transcripts.bed'))\n",
    "print(len(genes))\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t3214481\t3671498\tXkr4\t.\t-\tprotein_coding\t3214481\t3671498\t1\n",
      " chr1\t3999556\t4409241\tRp1\t.\t-\tprotein_coding\t3999556\t4409241\t3\n",
      " chr1\t4490930\t4496413\tSox17\t.\t-\tprotein_coding\t4490930\t4497354\t4\n",
      " chr1\t4773205\t4785710\tMrpl15\t.\t-\tprotein_coding\t4773205\t4785739\t3\n",
      " chr1\t4807822\t4846739\tLypla1\t.\t+\tprotein_coding\t4807822\t4846739\t2\n",
      " chr1\t4807891\t4886770\tRP24-426M1.3\t.\t+\tprotein_coding\t4807891\t4886770\t1\n",
      " chr1\t4807895\t4845174\tLypla1\t.\t+\tprotein_coding\t4807895\t4845352\t2\n",
      " chr1\t4857813\t4897905\tTcea1\t.\t+\tprotein_coding\t4857813\t4897909\t2\n",
      " chr1\t4909575\t5019539\tRgs20\t.\t-\tprotein_coding\t4909575\t5070285\t3\n",
      " chr1\t5070017\t5162340\tAtp6v1h\t.\t+\tprotein_coding\t5070017\t5162529\t3\n",
      " "
     ]
    }
   ],
   "source": [
    "coding = []\n",
    "\n",
    "for x in genes:\n",
    "    if x[6] == 'protein_coding':\n",
    "        coding.append(x)\n",
    "\n",
    "coding = BedTool(coding)\n",
    "coding = coding.groupby(g = [4,1], c = [2,3,4], o = ['min', 'max', 'count'], full=True)\n",
    "coding.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "rows = {'Gene': [], 'Chromosome': [], 'Start': [], 'End': []}\n",
    "\n",
    "for x in coding:\n",
    "    rows['Gene'].append(x[3])\n",
    "    rows['Chromosome'].append(x[0])\n",
    "    rows['Start'].append(int(x[1]))\n",
    "    rows['End'].append(int(x[2]))\n",
    "    \n",
    "logging.info(f\"Extracted coding genes to dict\")\n",
    "\n",
    "def Gene_count(ds, rows, res):\n",
    "    i = 0\n",
    "    count_dict = collections.OrderedDict()\n",
    "    \n",
    "    frags = ds.ca['fragments']\n",
    "    frags = [strFrags_to_list(x) for x in frags]\n",
    "    \n",
    "    logging.info(f\"extracted fragments\")\n",
    "    \n",
    "    for cell in ds.ca['cell_id']:\n",
    "        counts = {}\n",
    "\n",
    "        for _frag in frags:\n",
    "            for x in rows['Gene']:\n",
    "\n",
    "                if _frag[0] == rows['Chromosome'][rows['Gene'] == x]:\n",
    "\n",
    "                    st = rows['Start'][rows['Gene'] == x]\n",
    "                    en = rows['End'][rows['Gene'] == x]\n",
    "                    if np.logical_and(_frag[1] > st, _frag[2] < en) or np.logical_and(_frag[1] < en, _frag[2] > en) or np.logical_and(_frag[1] < st, _frag[2] > st):\n",
    "                        if x not in counts.keys():\n",
    "                            counts[x] = 1\n",
    "                        else:\n",
    "                            counts[x] += 1    \n",
    "            i += 1\n",
    "            res[cell] = counts\n",
    "            \n",
    "        if i%100 == 0:\n",
    "            logging.info(f\"Counted {i} cells\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'<class 'loompy.attribute_manager.AttributeManager'>' object has no attribute 'cell_ID'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/anaconda3/envs/chromograph/lib/python3.7/site-packages/loompy/attribute_manager.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m    114\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 115\u001b[0;31m                         \u001b[0mvals\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__dict__\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"storage\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    116\u001b[0m                         \u001b[0;32mif\u001b[0m \u001b[0mvals\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'cell_ID'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-24-bb2e9c158bf7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mfrags\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mca\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'fragments'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mfrags\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mstrFrags_to_list\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfrags\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mchunks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mca\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'cell_ID'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfrags\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/envs/chromograph/lib/python3.7/site-packages/loompy/attribute_manager.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, thing)\u001b[0m\n\u001b[1;32m    100\u001b[0m                                 \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    101\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 102\u001b[0;31m                         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mthing\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    103\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    104\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0m__getattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndarray\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/chromograph/lib/python3.7/site-packages/loompy/attribute_manager.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m    121\u001b[0m                         \u001b[0;32mreturn\u001b[0m \u001b[0mvals\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    122\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 123\u001b[0;31m                         \u001b[0;32mraise\u001b[0m \u001b[0mAttributeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"'{type(self)}' object has no attribute '{name}'\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    124\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    125\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0m__setitem__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mval\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndarray\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: '<class 'loompy.attribute_manager.AttributeManager'>' object has no attribute 'cell_ID'"
     ]
    }
   ],
   "source": [
    "vis = ['Neurod6', 'Gad2', 'Reln', 'Otx2', 'Olig2', 'Gfap', 'Igfbp2', 'Aqp4']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "keep = [x in vis for x in rows['Gene']]\n",
    "np.array(rows['Gene'])[k]\n",
    "\n",
    "short = {k: np.array(rows[k])[keep] for k in rows.keys()}\n",
    "short"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/camiel/anaconda3/envs/chromograph/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3326, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-7-159160c15401>\", line 1, in <module>\n",
      "    counts = Gene_count(ds, rows)\n",
      "  File \"<ipython-input-6-27c0b0876f68>\", line 21, in Gene_count\n",
      "    if _frag[0] == rows['Chromosome'][x]:\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/camiel/anaconda3/envs/chromograph/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2040, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'KeyboardInterrupt' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/camiel/anaconda3/envs/chromograph/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/home/camiel/anaconda3/envs/chromograph/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 319, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/home/camiel/anaconda3/envs/chromograph/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 353, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/home/camiel/anaconda3/envs/chromograph/lib/python3.7/inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/home/camiel/anaconda3/envs/chromograph/lib/python3.7/inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/home/camiel/anaconda3/envs/chromograph/lib/python3.7/inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/home/camiel/anaconda3/envs/chromograph/lib/python3.7/inspect.py\", line 739, in getmodule\n",
      "    f = getabsfile(module)\n",
      "  File \"/home/camiel/anaconda3/envs/chromograph/lib/python3.7/inspect.py\", line 708, in getabsfile\n",
      "    _filename = getsourcefile(object) or getfile(object)\n",
      "  File \"/home/camiel/anaconda3/envs/chromograph/lib/python3.7/inspect.py\", line 693, in getsourcefile\n",
      "    if os.path.exists(filename):\n",
      "  File \"/home/camiel/anaconda3/envs/chromograph/lib/python3.7/genericpath.py\", line 19, in exists\n",
      "    os.stat(path)\n",
      "KeyboardInterrupt\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m"
     ]
    }
   ],
   "source": [
    "res = {}\n",
    "counts = Gene_count(ds, short, res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ".\n",
      "1000\n",
      "2000\n",
      "3000\n",
      "4000\n",
      "5000\n",
      "6000\n",
      "7000\n",
      "8000\n",
      "9000\n",
      "10000\n",
      "11000\n",
      "12000\n",
      "13000\n",
      "14000\n",
      "15000\n",
      "16000\n",
      "17000\n",
      "18000\n",
      "19000\n",
      "20000\n",
      "21000\n",
      "22000\n",
      "23000\n",
      "24000\n",
      "25000\n",
      ".\n",
      "1000\n",
      "2000\n",
      "3000\n",
      "4000\n",
      "5000\n",
      "6000\n",
      "7000\n",
      "8000\n",
      "9000\n",
      "10000\n",
      "11000\n",
      "12000\n",
      "13000\n",
      "14000\n",
      "15000\n",
      "16000\n",
      "17000\n",
      "18000\n",
      "19000\n",
      "20000\n",
      "21000\n",
      "22000\n",
      "23000\n",
      "24000\n",
      "25000\n",
      ".\n",
      "1000\n",
      "2000\n",
      "3000\n",
      "4000\n",
      "5000\n",
      "6000\n",
      "7000\n",
      "8000\n",
      "9000\n",
      "10000\n",
      "11000\n",
      "12000\n",
      "13000\n",
      "14000\n",
      "15000\n",
      "16000\n",
      "17000\n",
      "18000\n",
      "19000\n",
      "20000\n",
      "21000\n",
      "22000\n",
      "23000\n",
      "24000\n",
      "25000\n",
      ".\n",
      "1000\n",
      "2000\n",
      "3000\n",
      "4000\n",
      "5000\n",
      "6000\n",
      "7000\n",
      "8000\n",
      "9000\n",
      "10000\n",
      "11000\n",
      "12000\n",
      "13000\n",
      "14000\n",
      "15000\n",
      "16000\n",
      "17000\n",
      "18000\n",
      "19000\n",
      "20000\n",
      "21000\n",
      "22000\n",
      "23000\n",
      "24000\n",
      "25000\n",
      ".\n",
      "1000\n",
      "2000\n",
      "3000\n",
      "4000\n",
      "5000\n",
      "6000\n",
      "7000\n",
      "8000\n",
      "9000\n",
      "10000\n",
      "11000\n",
      "12000\n",
      "13000\n",
      "14000\n",
      "15000\n",
      "16000\n",
      "17000\n",
      "18000\n",
      "19000\n",
      "20000\n",
      "21000\n",
      "22000\n",
      "23000\n",
      "24000\n",
      "25000\n",
      ".\n",
      "1000\n",
      "2000\n",
      "3000\n",
      "4000\n",
      "5000\n",
      "6000\n",
      "7000\n",
      "8000\n",
      "9000\n",
      "10000\n",
      "11000\n",
      "12000\n",
      "13000\n",
      "14000\n",
      "15000\n",
      "16000\n",
      "17000\n",
      "18000\n",
      "19000\n",
      "20000\n",
      "21000\n",
      "22000\n",
      "23000\n",
      "24000\n",
      "25000\n",
      ".\n",
      "1000\n",
      "2000\n",
      "3000\n",
      "4000\n",
      "5000\n",
      "6000\n",
      "7000\n",
      "8000\n",
      "9000\n",
      "10000\n",
      "11000\n",
      "12000\n",
      "13000\n",
      "14000\n",
      "15000\n",
      "16000\n",
      "17000\n",
      "18000\n",
      "19000\n",
      "20000\n",
      "21000\n",
      "22000\n",
      "23000\n",
      "24000\n",
      "25000\n",
      ".\n",
      "1000\n",
      "2000\n",
      "3000\n",
      "4000\n",
      "5000\n",
      "6000\n",
      "7000\n",
      "8000\n",
      "9000\n",
      "10000\n",
      "11000\n",
      "12000\n",
      "13000\n",
      "14000\n",
      "15000\n",
      "16000\n",
      "17000\n",
      "18000\n",
      "19000\n",
      "20000\n",
      "21000\n",
      "22000\n",
      "23000\n",
      "24000\n",
      "25000\n",
      "2min 58s ± 1.28 s per loop (mean ± std. dev. of 7 runs, 1 loop each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "\n",
    "cell = '_cell_0'\n",
    "\n",
    "count_dict = collections.OrderedDict()\n",
    "counts = {}\n",
    "\n",
    "frags = ds.ca['fragments'][ds.ca['cell_id']==cell]\n",
    "frags = strFrags_to_list(frags[0])\n",
    "\n",
    "i = 0\n",
    "for _frag in frags:\n",
    "    for x in rows['Gene']:\n",
    "\n",
    "        if _frag[0] == rows['Chromosome'][rows['Gene'] == x]:\n",
    "            \n",
    "            st = rows['Start'][rows['Gene'] == x]\n",
    "            en = rows['End'][rows['Gene'] == x]\n",
    "            if np.logical_and(_frag[1] > st, _frag[2] < en) or np.logical_and(_frag[1] < en, _frag[2] > en) or np.logical_and(_frag[1] < st, _frag[2] > st):\n",
    "                if x not in counts.keys():\n",
    "                    counts[x] = 1\n",
    "                else:\n",
    "                    counts[x] += 1\n",
    "    i += 1\n",
    "    \n",
    "    if i%1000 == 0:\n",
    "        print(i)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def count_fragments(frag_dict, barcodes, bsize):\n",
    "#     '''\n",
    "#     '''\n",
    "    \n",
    "#     Count_dict = collections.OrderedDict()\n",
    "\n",
    "#     for bar in barcodes:    \n",
    "#         if bar in frag_dict:\n",
    "#             frags = frag_dict[bar]\n",
    "#             counts = {}\n",
    "#             for _frag in frags:\n",
    "\n",
    "#                 # If a fragment spans two bins we count it twice\n",
    "#                 for x in set([int(_frag[1]/bsize)*bsize+1, int(_frag[2]/bsize)*bsize+1]):\n",
    "#                     k = (_frag[0], x, x + bsize - 1)\n",
    "#                     if k not in counts.keys():\n",
    "#                         counts[k] = 1\n",
    "#                     else:\n",
    "#                         counts[k] += 1\n",
    "#             Count_dict[bar] = counts\n",
    "#         else:\n",
    "#             continue\n",
    "    \n",
    "#     return Count_dict;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19883\n"
     ]
    }
   ],
   "source": [
    "feature = []\n",
    "for x in coding:\n",
    "    feature.append(x[3])\n",
    "    \n",
    "print(len(np.unique(feature)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_fragments(frag_dict, barcodes, bsize):\n",
    "    '''\n",
    "    '''\n",
    "    \n",
    "    Count_dict = collections.OrderedDict()\n",
    "\n",
    "    for bar in barcodes:    \n",
    "        if bar in frag_dict:\n",
    "            frags = frag_dict[bar]\n",
    "            counts = {}\n",
    "            for _frag in frags:\n",
    "\n",
    "                # If a fragment spans two bins we count it twice\n",
    "                for x in set([int(_frag[1]/bsize)*bsize+1, int(_frag[2]/bsize)*bsize+1]):\n",
    "                    k = (_frag[0], x, x + bsize - 1)\n",
    "                    if k not in counts.keys():\n",
    "                        counts[k] = 1\n",
    "                    else:\n",
    "                        counts[k] += 1\n",
    "            Count_dict[bar] = counts\n",
    "        else:\n",
    "            continue\n",
    "    \n",
    "    return Count_dict;\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Chromograph (py3.7)",
   "language": "python",
   "name": "chromograph"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
